### PR TITLE
 Longevity test for MV and SI

### PR DESCRIPTION
**Pull request by juliayakovlev**
Few changes:
1. added 4 new user profiles for MV&SI test
2. added new MV&SI test yaml
3. change test_custom_time so that it can run a few commands in the prepare sector and it may be user profiles
4. fix nemesis - able to fetch test keyspaces from the DB and perform actions on keyspaces, that different from "keyspace1"

### CODE REVIEW

File: data_dir/longevity_mv_prefilled_profile.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+keyspace: mview
```
Comment (roydahan) [inline data_dir/longevity_mv_prefilled_profile.yaml]: The name of the profile isn't clear.
I don't understand what is prefiiled and what is the diff between this and the other one.

I suggest you call it according to its properties, something like c-s_profile_2_views_7_queries.yaml

File: data_dir/longevity_mv_profile.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
+keyspace: mview
```
Comment (roydahan) [inline data_dir/longevity_mv_profile.yaml]: same as above, if we change to meaningful names.

File: data_dir/longevity_si_prefilled_profile.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+keyspace: sec_index
```
Comment (roydahan) [inline data_dir/longevity_si_prefilled_profile.yaml]: same as above.

File: data_dir/longevity_si_profile.yaml [hunk 4]
```diff
@@ -0,0 +1,74 @@
+keyspace: sec_index
```
Comment (roydahan) [inline data_dir/longevity_si_profile.yaml]: same as above

File: sdcm/cluster.py [hunk 5]
```diff
@@ -1683,35 +1683,51 @@ def get_scylla_version(self):
...
         :return: Whether all nodes reached that threshold or not.
         """
-        tcpdump = self.params.get('tcpdump')
-        cfstats = [node.get_cfstats(keyspace, tcpdump)[key] for node in self.nodes]
-        self.log.debug("Current cfstats: %s" % cfstats)
+        if not keyspaces:
+            keyspaces = self.get_test_keyspaces()
+            # out = self.run_cqlsh(node=self.nodes[0], cql_cmd='select keyspace_name from system_schema.keyspaces')
+            # keyspaces = [ks.strip() for ks in out.split('\n')[3:-3] if 'system' not in ks]
+
```
Comment (roydahan) [inline sdcm/cluster.py]: Why were these comments left here?

File: sdcm/cluster.py [hunk 6]
```diff
@@ -1683,35 +1683,51 @@ def get_scylla_version(self):
...
+        cfstats = []
+        for keyspace_name in keyspaces:
+            ks_stats = [node.get_cfstats(keyspace_name)[key] for node in self.nodes]
```
Comment (roydahan) [inline sdcm/cluster.py]: I know we've done it till now, but is it really necessary to run cfstats on all nodes?

File: sdcm/cluster.py [hunk 7]
```diff
@@ -1683,35 +1683,51 @@ def get_scylla_version(self):
...
+            cfstats.append(min(ks_stats))
+            self.log.debug("Current cfstats for %s keyspace: %s" % (keyspace_name, ks_stats))
```
Comment (roydahan) [inline sdcm/cluster.py]: Let's print in this log also the threshold we are waiting for, so it'll east to detect issues.

File: sdcm/cluster.py [hunk 8]
```diff
@@ -1683,35 +1683,51 @@ def get_scylla_version(self):
...
-
     def wait_total_space_used_per_node(self, size=None, keyspace='keyspace1'):
         if size is None:
             size = int(self.params.get('space_node_threshold'))
-        self.wait_cfstat_reached_threshold('Space used (total)', size, keyspace)
+        if size:
+            if keyspace and not isinstance(keyspace, list):
+                keyspace = [keyspace]
+            key = 'Space used (total)'
+            wait.wait_for(func=self.cfstat_reached_threshold, step=10,
```
Comment (roydahan) [inline sdcm/cluster.py]: maybe we should set a timeout here.

File: sdcm/cluster.py [hunk 9]
```diff
@@ -2001,8 +2017,8 @@ def kill_stress_thread(self):
             loader.remoter.run(cmd='rm -rf %s' % kill_script_dir)
         kill_script.remove()
 
-    @staticmethod
-    def _parse_cs_summary(lines):
+    # @staticmethod
+    def _parse_cs_summary(self, lines):
```
Comment (roydahan) [inline sdcm/cluster.py]: Why the change?
Are you sure it doesn't affect anything?

File: sdcm/cluster.py [hunk 10]
```diff
@@ -2013,6 +2029,8 @@ def _parse_cs_summary(lines):
 
         for line in lines:
             line.strip()
+            if not line:
+                continue
```
Comment (roydahan) [inline sdcm/cluster.py]: Why needed?

File: tests/longevity-mv-si-2days.yaml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+nemesis_interval: 30
+nemesis_during_prepare: 'true'
+user_prefix: 'julia-longevity-mv-si-2d'
+failure_post_behavior: keep
+space_node_threshold: 325
+ip_ssh_connections: 'public'
+#When the load is too heavy for one lader when using MULTI-KEYSPACES, the load is spreaded evenly across the loaders (round_robin)
+round_robin: 'false'
+experimental: 'true'
+keyspace_num: 2
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: why?

File: tests/longevity-mv-si-2days.yaml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                 ]
+n_db_nodes: 3
+n_loaders: 2
+n_monitor_nodes: 1
+nemesis_class_name: 'AllMonkey'
+nemesis_interval: 30
+nemesis_during_prepare: 'true'
+user_prefix: 'julia-longevity-mv-si-2d'
+failure_post_behavior: keep
+space_node_threshold: 325
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: I think it should be higher

File: tests/longevity-mv-si-2days.yaml [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+stress_read_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_prefilled_profile.yaml ops'(insert=21,mv_p_read1=1,mv_p_read2=1,mv_p_read3=1,mv_p_read4=1,mv_p_read5=1,mv_p_read6=1,mv_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+                  "cassandra-stress user profile=/tmp/longevity_si_prefilled_profile.yaml ops'(insert=35,si_p_read1=1,si_p_read2=1,si_p_read3=1,si_p_read4=1,si_p_read5=1,si_p_read6=1,si_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+                 ]
+n_db_nodes: 3
+n_loaders: 2
+n_monitor_nodes: 1
+nemesis_class_name: 'AllMonkey'
+nemesis_interval: 30
+nemesis_during_prepare: 'true'
+user_prefix: 'julia-longevity-mv-si-2d'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: change to longevity-mv-si-2d-VERSION

File: tests/longevity-mv-si-2days.yaml [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            ]
+stress_read_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_prefilled_profile.yaml ops'(insert=21,mv_p_read1=1,mv_p_read2=1,mv_p_read3=1,mv_p_read4=1,mv_p_read5=1,mv_p_read6=1,mv_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+                  "cassandra-stress user profile=/tmp/longevity_si_prefilled_profile.yaml ops'(insert=35,si_p_read1=1,si_p_read2=1,si_p_read3=1,si_p_read4=1,si_p_read5=1,si_p_read6=1,si_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+                 ]
+n_db_nodes: 3
+n_loaders: 2
+n_monitor_nodes: 1
+nemesis_class_name: 'AllMonkey'
+nemesis_interval: 30
+nemesis_during_prepare: 'true'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: You don't have prepare, so why true?

File: tests/longevity-mv-si-2days.yaml [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+stress_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_profile.yaml ops'(insert=15,read1=1,read2=1,read3=1,read4=1,read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+             "cassandra-stress user profile=/tmp/longevity_si_profile.yaml ops'(insert=25,si_read1=1,si_read2=1,si_read3=1,si_read4=1,si_read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+            ]
+stress_read_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_prefilled_profile.yaml ops'(insert=21,mv_p_read1=1,mv_p_read2=1,mv_p_read3=1,mv_p_read4=1,mv_p_read5=1,mv_p_read6=1,mv_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+                  "cassandra-stress user profile=/tmp/longevity_si_prefilled_profile.yaml ops'(insert=35,si_p_read1=1,si_p_read2=1,si_p_read3=1,si_p_read4=1,si_p_read5=1,si_p_read6=1,si_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+                 ]
+n_db_nodes: 3
+n_loaders: 2
+n_monitor_nodes: 1
+nemesis_class_name: 'AllMonkey'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: Shouldn't be ChaosMoneky?
What are the differences?

File: tests/longevity-mv-si-2days.yaml [hunk 16]
```diff
@@ -0,0 +1,81 @@
+test_duration: 2880
+stress_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_profile.yaml ops'(insert=15,read1=1,read2=1,read3=1,read4=1,read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+             "cassandra-stress user profile=/tmp/longevity_si_profile.yaml ops'(insert=25,si_read1=1,si_read2=1,si_read3=1,si_read4=1,si_read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+            ]
+stress_read_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_prefilled_profile.yaml ops'(insert=21,mv_p_read1=1,mv_p_read2=1,mv_p_read3=1,mv_p_read4=1,mv_p_read5=1,mv_p_read6=1,mv_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+                  "cassandra-stress user profile=/tmp/longevity_si_prefilled_profile.yaml ops'(insert=35,si_p_read1=1,si_p_read2=1,si_p_read3=1,si_p_read4=1,si_p_read5=1,si_p_read6=1,si_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+                 ]
+n_db_nodes: 3
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: Should be 5 (we temporarily wanted to run with 3).

File: tests/longevity-mv-si-2days.yaml [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+n_loaders: 2
+n_monitor_nodes: 1
+nemesis_class_name: 'AllMonkey'
+nemesis_interval: 30
+nemesis_during_prepare: 'true'
+user_prefix: 'julia-longevity-mv-si-2d'
+failure_post_behavior: keep
+space_node_threshold: 325
+ip_ssh_connections: 'public'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: Change to private

File: tests/longevity-mv-si-2days.yaml [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        gce_root_disk_type_monitor: 'pd-standard'
+        gce_root_disk_size_monitor: 50
+        gce_n_local_ssd_disk_monitor: 0
+        scylla_repo: 'REPO_FILE_PATH'
+        us_east_1:
+          gce_datacenter: 'us-east1-b'
+    aws: !mux
+        cluster_backend: 'aws'
+        user_credentials_path: '~/.ssh/scylla-qa-ec2'
+        instance_type_loader: 'c4.4xlarge'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: change to c4.large

File: tests/longevity-mv-si-2days.yaml [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+        instance_type_monitor: 'c4.2xlarge'
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: change to t2.small

File: tests/longevity-mv-si-2days.yaml [hunk 20]
```diff
@@ -0,0 +1,81 @@
+test_duration: 2880
+stress_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_profile.yaml ops'(insert=15,read1=1,read2=1,read3=1,read4=1,read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: remove the -errors ignore
threads=10 is enough load?

File: tests/longevity-mv-si-2days.yaml [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+             "cassandra-stress user profile=/tmp/longevity_si_profile.yaml ops'(insert=25,si_read1=1,si_read2=1,si_read3=1,si_read4=1,si_read5=1)' cl=QUORUM n=10000000 -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
+            ]
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: Same as above

File: tests/longevity-mv-si-2days.yaml [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+stress_read_cmd: ["cassandra-stress user profile=/tmp/longevity_mv_prefilled_profile.yaml ops'(insert=21,mv_p_read1=1,mv_p_read2=1,mv_p_read3=1,mv_p_read4=1,mv_p_read5=1,mv_p_read6=1,mv_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore",
+                  "cassandra-stress user profile=/tmp/longevity_si_prefilled_profile.yaml ops'(insert=35,si_p_read1=1,si_p_read2=1,si_p_read3=1,si_p_read4=1,si_p_read5=1,si_p_read6=1,si_p_read7=1)' cl=QUORUM duration=2880m -port jmx=6868 -mode cql3 native -rate threads=10 -errors ignore"
```
Comment (roydahan) [inline tests/longevity-mv-si-2days.yaml]: same

File: longevity_test.py [hunk 23]
```diff
@@ -127,24 +138,27 @@ def test_custom_time(self):
...
-        else:
-                params = {'keyspace_num': keyspace_num}
-                self._run_all_stress_cmds(stress_queue, params)
+            # The old method when we run all stress_cmds for all keyspace on the same loader
+            else:
+                    params = {'keyspace_num': keyspace_num, 'stress_cmd': stress_cmd}
+                    self._run_all_stress_cmds(stress_queue, params)
 
         # Check if we shall wait for total_used_space or if nemesis wasn't started
         if not prepare_write_cmd or self.params.get('nemesis_during_prepare', default='true').lower() == 'false':
```
Comment (roydahan) [inline longevity_test.py]: I think you can remove the sleep(600) below and use the fixed wait_for_total_space now.

File: tests/longevity-mv-si-2days.yaml (hunk 20 above)
Comment (juliayakovlev) [reply tests/longevity-mv-si-2days.yaml]: I put "-errors ignore" by Shlomi's request

File: sdcm/cluster.py (hunk 10 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: Because if there is empty line in the output, parsing fails.

File: sdcm/cluster.py (hunk 9 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: I did it to be able to print debug info and forgot to revert before commit. Thanks

File: sdcm/cluster.py (hunk 8 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: Put to 300 sec. Is it ok?

File: sdcm/cluster.py (hunk 6 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: You asked to check on every node

File: sdcm/cluster.py (hunk 5 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: Forgot to remove. Thanks

File: data_dir/longevity_mv_prefilled_profile.yaml (hunk 1 above)
Comment (juliayakovlev) [reply data_dir/longevity_mv_prefilled_profile.yaml]: We need to discuss about the names

**Comment by juliayakovlev** [comment]
@roydahan @bentsi  Please, review

File: sdcm/cluster_aws.py [hunk 24]
```diff
@@ -328,7 +328,7 @@ def restart(self):
         # Restart in AWS will be a Stop and Start of an instance.
         # When using storage optimized instances like i2 or i3, the data on disk is deleted upon STOP. therefore, we
         # need to setup the instance and treat it as a new instance.
-        if ('i2' or 'i3') in self._instance.instance_type:
+        if any(ss in self._instance.instance_type for ss in ['i3', 'i2']):
```
Comment (bentsi) [inline sdcm/cluster_aws.py]: Nice catch and solution!!!

File: sdcm/cluster.py [hunk 25]
```diff
@@ -1683,35 +1683,53 @@ def get_scylla_version(self):
                     else:
                         self.log.error("Unknown ScyllaDB version")
 
-    def cfstat_reached_threshold(self, key, threshold, keyspace):
+    def run_cqlsh(self, node, cql_cmd, timeout=60, verbose=True):
```
Comment (bentsi) [inline sdcm/cluster.py]: Good! all places that uses `cqlsh` should use this method

File: longevity_test.py [hunk 26]
```diff
@@ -127,46 +138,37 @@ def test_custom_time(self):
...
+                self._run_all_stress_cmds(verify_queue, params={'stress_cmd': prepare_verify_cmd,
+                                                                'keyspace_num': keyspace_num})
 
                 for stress in verify_queue:
                     self.verify_stress_thread(queue=stress)
 
-        # Stress: Same as in prepare_write - allow the load to be spread across all loaders when using MULTI-KEYSPACES
-        if keyspace_num > 1 and self.params.get('round_robin', default='').lower() == 'true':
+        stress_cmd = self.params.get('stress_cmd', default=None)
+        if stress_cmd:
```
Comment (bentsi) [inline longevity_test.py]: good!

File: sdcm/cluster.py (hunk 10 above)
Comment (roydahan) [reply sdcm/cluster.py]: It's the same as what written few lines below:
if line == '':
                 continue

**Comment by bentsi** [comment]
Merging this to master, @juliayakovlev we will need to backport to branch-3.0 when some some testing will be done on master

**Comment by roydahan** [comment]
@bentsi, are you backporting this?

**Comment by bentsi** [comment]
Backported to `branch-3.0`
